community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/executionplan/StandardInternalExecutionResult.scala

```diff
@@ -97,10 +97,14 @@ abstract class StandardInternalExecutionResult(context: QueryContext, runtime: R
 
   override def accept[E <: Exception](visitor: ResultVisitor[E]): Unit = {
     accept(new QueryResultVisitor[E] {
+      val names = fieldNames()
       override def visit(record: QueryResult.Record): Boolean = {
-        val row = new MapBasedRow
-        row.map = fieldNames().zip(record.fields().map(context.asObject)).toMap
-        visitor.visit(row)
+        val fields = record.fields()
+        val mapData = new mutable.AnyRefMap[String, Any](names.length)
```

Comment:
We do the same thing in `PipeExecutionResult`, it should be just as inefficient there. Can we apply the same fix there?

Reply:
Good point. I've added a new commit for this. I also profiled it with `MATCH (n) RETURN n` and in that case the previous code used 40% of the result generation time to convert the results to external types with this method, while after the commit the code uses 20% of the time. So, as before, we double the performance of this specific algorithm.